osquery/events/darwin/openbsm.cpp

```diff
@@ -41,54 +42,83 @@ void OpenBSMEventPublisher::tearDown() {
   }
 }
 
+void OpenBSMEventPublisher::acquireMessages() {
+  auto buffer = static_cast<unsigned char*>(nullptr);
+  auto reclen = au_read_rec(audit_pipe_, &buffer);
+  if (reclen <= 0) {
+    return;
+  }
+
+  // We'll use these to dequeue below.
+  tokenstr_t tok;
+  std::vector<tokenstr_t> tokens{};
+
+  auto event_id = 0;
+  auto bytesread = 0;
+  while (bytesread < reclen) {
+    if (au_fetch_tok(&tok, buffer + bytesread, reclen - bytesread) == -1) {
+      break;
+    }
+    switch (tok.id) {
+    case AUT_HEADER32:
+      event_id = tok.tt.hdr32_ex.e_type;
+      break;
+    case AUT_HEADER32_EX:
+      event_id = tok.tt.hdr32_ex.e_type;
+      break;
+    case AUT_HEADER64:
+      event_id = tok.tt.hdr64.e_type;
+      break;
+    case AUT_HEADER64_EX:
+      event_id = tok.tt.hdr64_ex.e_type;
+      break;
+    }
+    tokens.push_back(tok);
+    bytesread += tok.len;
+  }
+
+  // We probably don't need a lambda here but it's useful to put debug
+  // lines in to validate destruction.
+  std::shared_ptr<unsigned char> sp_buffer(buffer,
+                                           [](unsigned char* p) { delete p; });
+
+  auto ec = createEventContext();
+  ec->event_id = event_id;
+  ec->tokens = tokens;
+  ec->buffer = sp_buffer;
+  fire(ec);
+}
+
 Status OpenBSMEventPublisher::run() {
   if (audit_pipe_ == nullptr) {
     return Status(1, "No open audit_pipe");
   }
-  tokenstr_t tok;
-  auto reclen = 0;
-  auto bytesread = 0;
-  auto event_id = 0;
-  auto buffer = static_cast<unsigned char*>(nullptr);
-  std::vector<tokenstr_t> tokens{};
 
-  while (!isEnding() && (reclen = au_read_rec(audit_pipe_, &buffer)) != -1) {
-    bytesread = 0;
+  fd_set fdset;
+  FD_ZERO(&fdset);
+  FD_SET(fileno(audit_pipe_), &fdset);
+  struct timeval timeout;
+  timeout.tv_sec = 0;
+  timeout.tv_usec = 20000;
```

Comment:
I believe timeout.tv_usec = 20000 too small wait time (its in micro seconds).
it will induce 50 context switches in a second in idle scenario

Reply:
You are correct, this is a very short time to wait for data. I found that since the subsequent call to dequeue data from audit is blocking and because it only dequeues a single event, we need to run this loop quickly when data is available.

When data is not available this method will return to https://github.com/osquery/osquery/blob/04896c85cd14c7ea8dc4e8fda6aaff4e7001d198/osquery/events/events.cpp#L798 which will apply a longer 200 millisecond wait.

Reply:
oh! yes there is a break.

if we are going for 200ms wait and there are lot of events (specially when we start capturing more types of event) getting generated then internal kernel queue(max lenght 1024 evenst) might get full and that will result into event loss.

Reply:
Correct, another way to implement this polling is to never let it reach the return (and prevent falling through to the catch-all pause above).

I propose we start here and then iterate in a follow up PR.